core/src/main/java/com/linecorp/armeria/client/SessionProtocolNegotiationCache.java

```diff
@@ -160,24 +155,41 @@ private static CacheEntry getOrCreate(String key) {
         }
     }
 
-    private static String key(SocketAddress remoteAddress) {
+    @VisibleForTesting
+    static String key(Endpoint endpoint) {
+        checkArgument(endpoint.hasPort() || endpoint.isDomainSocket(),
+                      "endpoint: %s (expected: has a port or is a domain socket address)", endpoint);
+
+        // It's okay to create the key using endpoint.authority():
+        // - If the endpoint has an IP address without host name, the IP address is used in the key to store
+        //   and retrieve the value.
+        // - If the endpoint has an IP address with host name, the host name is used in the key to store
+        //   and retrieve the value.
+        // - If the endpoint has host name only, the host name is used in the key to store and retrieve
+        //   the value.
+        // - If the endpoint is a domain socket address, the encoded path is used in the key to store and
+        //   retrieve the value.
+        return endpoint.authority();
```

Comment:
Question) Is there no need to use a single key format for `endpoint` and `socketAddress`?

e.g. Assuming ipv6

e.g.
1. `SessionProtocolNegotiationCache#setUnsupported` is only set for `SocketAddress` type
2. The key for `SocketAddress` is ("1:2:3:4...:port")
3. The key format for `Endpoint` is `[1:2:3:4...]:port`
4. The cache check for `Endpoint` is never hit for ipv6